### PR TITLE
LibreSSL support

### PR DESCRIPTION
**Pull request by reaperhulk**
This PR replaces #1673, #1674, and #1636.

Huge thanks to @Anchorcat for the hard work. I've adapted the cryptodev and @sp1l's EGD work to be a conditional binding. (Be aware that with EGD downstream consumers like pyOpenSSL may not work. I haven't tested.).

I've also added a libre builder to jenkins so if we choose to merge this we'll need to rebase all other PRs (and, conversely, if we choose to not merge this I need to remove that builder).

### CODE REVIEW

**Comment by alex** [comment]
This PR looks on-ball to me. Will wait to see how the tests look.

**Comment by reaperhulk** [comment]
I'll retest jenkins once it finishes failing. We can't run with `-Werror` due to deprecations around `CRYPTO_dbg_*` so I had to update the jenkins config.

**Comment by reaperhulk** [comment]
retest this please

**Comment by alex** [comment]
@reaperhulk besides the final reviewer, are there blockers infrastructure wise to merging this?

**Comment by reaperhulk** [comment]
Once we're ready to merge we need to add the libre config to the cryptography-master job (it's only on pr-experimental at the moment). That's the only infra item left. Do we want anyone else to weigh in or is this ready?

File: src/cryptography/hazmat/bindings/openssl/ssl.py [hunk 1]
```diff
@@ -7,6 +7,14 @@
 INCLUDES = """
 #include <openssl/ssl.h>
 
+/* LibreSSL has removed support for compression, and with it the
+ * COMP_METHOD use in ssl.h. This is a hack to make the function types
+ * in this code match those in ssl.h.
+ */
+#ifdef LIBRESSL_VERSION_NUMBER
+#define COMP_METHOD void
+#endif
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Should compression related be in `CONDITIONAL`?

**Comment by alex** [comment]
Besides that one question, this LGTM.

File: src/cryptography/hazmat/bindings/openssl/ssl.py (hunk 1 above)
Comment (reaperhulk) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Yeah, although that starts to bleed into #1675 :grinning: 

I'll move this so we can merge today.

Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Sorry, move what?

On Thu, Feb 19, 2015 at 2:53 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/bindings/openssl/ssl.py
> <https://github.com/pyca/cryptography/pull/1679#discussion_r25019164>:
>
> > @@ -7,6 +7,14 @@
> >  INCLUDES = """
> >  #include <openssl/ssl.h>
> >
> > +/* LibreSSL has removed support for compression, and with it the
> > + * COMP_METHOD use in ssl.h. This is a hack to make the function types
> > + * in this code match those in ssl.h.
> > + */
> > +#ifdef LIBRESSL_VERSION_NUMBER
> > +#define COMP_METHOD void
> > +#endif
>
> Yeah, although that starts to bleed into #1675
> <https://github.com/pyca/cryptography/pull/1675> [image: :grinning:]
>
> I'll move this so we can merge today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1679/files#r25019164>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (reaperhulk) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Move the COMP_METHOD definition.

**Comment by AnchorCat** [comment]
Thanks! I've tested the latest master with tox on my OpenBSD -current system, and all tests pass with both python2.7 and python3.4 (just thought I'd confirm in case there were differences with LibreSSL portable).

Out of curiosity, I ran the pyOpenSSL test suite against this version, and got 3 failures and 6 errors (out of 397 tests). I've uploaded the [failed test output](http://steven.beta.anchortrove.com/pyopenssl-test.txt) in case anyone is interested in the details. I may look at fixing those if I have time, but pyOpenSSL is working well enough for my real-world use case (the `python-glanceclient` utility).

**Pull request merged**